common/src/com/thoughtworks/go/buildsession/ExecCommandExecutor.java

```diff
@@ -0,0 +1,122 @@
+        File workingDir = buildSession.resolveRelativeDir(command.getWorkingDirectory());
+        if (!workingDir.isDirectory()) {
+            String message = "Working directory \"" + workingDir.getAbsolutePath() + "\" is not a directory!";
+            LOG.error(message);
+            buildSession.println(message);
+            return false;
+        }
+
+        String cmd = command.getStringArg("command");
+        String[] args = command.getArrayArg("args");
+        CommandLine commandLine;
+
+        if (SystemUtil.isWindows()) {
+            commandLine = CommandLine.createCommandLine("cmd");
+            commandLine.withArg("/c");
+            commandLine.withArg(StringUtils.replace(cmd, "/", "\\"));
+        } else {
+            commandLine = CommandLine.createCommandLine(cmd);
+        }
+        commandLine.withWorkingDir(workingDir);
+        commandLine.withEnv(buildSession.getEnvs());
+        commandLine.withArgs(args);
+        for (SecretSubstitution secretSubstitution : buildSession.getSecretSubstitutions()) {
+            commandLine.withNonArgSecret(secretSubstitution);
+        }
+        return executeCommandLine(buildSession, commandLine) == 0;
+    }
+
+    private int executeCommandLine(final BuildSession buildSession, final CommandLine commandLine) {
+        ExecutorService executorService = Executors.newFixedThreadPool(2);
+        final AtomicInteger exitCode = new AtomicInteger(-1);
+        final CountDownLatch canceledOrDone = new CountDownLatch(1);
+
+        try {
+            final Future<?> executing = executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        exitCode.set(commandLine.run(buildSession.processOutputStreamConsumer(), null));
+                    } catch (CommandLineException e) {
+                        LOG.error("Command failed", e);
+                        String message = format("Error happened while attempting to execute '%s'. \nPlease make sure [%s] can be executed on this agent.\n", commandLine.toStringForDisplay(), commandLine.getExecutable());
+                        String path = System.getenv("PATH");
+                        buildSession.println(message);
+                        buildSession.println(format("[Debug Information] Environment variable PATH: %s", path));
+                        LOG.error(format("[Command Line] %s. Path: %s", message, path));
+                    } finally {
+                        canceledOrDone.countDown();
+                    }
+                }
+            });
+
+            Future<?> cancelMonitor = executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        buildSession.waitUntilCanceled();
+                        executing.cancel(true);
+                        new JavaSysMon().infanticide();
```

Comment:
There is another bonus for the new canceling implementation:  because it is implemented on exec command level, it can cleanly cancel build hangs at material updating stage. In current trunks implementation when `git fetch` hangs, user need kill process manually to get agent back.